src/cmd/ksh93/tests/modifiers.sh

```diff
@@ -0,0 +1,42 @@
+# These are the tests for various style of parameter expansion modifiers supported by ksh93
+
+# Check ${parameter:-word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+unset foo
+[[ ${foo:-bar} == bar ]]  || log_error  '${foo:-bar} not bar when foo is not set'
+
+unset foo
+[[ ${foo-bar} == bar ]]  || log_error  '${foo-bar} not bar when foo is not set'
+
+foo=""
+[[ ${foo-bar} == "" ]]  || log_error  '${foo-bar} not "" when foo is null'
+
+# Check ${parameter:=word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+unset foo
+[[ ${foo:=bar} == bar ]]  || log_error '${foo:=bar} not bar when foo is not set'
+
+unset foo
+[[ ${foo=bar} == bar ]]  || log_error '${foo=bar} not bar when foo is not set'
+
+foo=""
+[[ ${foo=bar} == "" ]]  || log_error  '${foo=bar} not "" when foo is null'
+
+# Check ${parameter:?word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+[[ $( $SHELL -c 'unset foo; print ${foo:?bar not set}' 2>&1 ) =~ "bar not set" ]]  || log_error '${foo:?bar} does not display error if foo is unset'
+
+[[ $( $SHELL -c 'unset foo; print ${foo?bar not set}' 2>&1 ) =~ "bar not set" ]]  || log_error '${foo?bar} does not display error if foo is unset'
+
+[[ $( $SHELL -c 'foo=""; print ${foo?bar}' 2>&1 ) == "" ]]  || log_error  '${foo=bar} not "" when foo is null'
+
+# When nothing is specified after :?, a default error message is printed
+unset foo
+[[ $( (print ${foo:?}) 2>&1) =~ "parameter not set" ]] || log_error 'Incorrect error message with ${foo?}'
+
+# Check ${parameter:+word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+unset foo
+[[ ${foo:+bar} == "" ]]  || log_error '${foo:+bar} not null when foo is not set'
+
+unset foo
+[[ ${foo+bar} == "" ]]  || log_error '${foo+bar} not null when foo is not set'
+
+foo="non-null value"
+[[ ${foo+bar} == "bar" ]]  || log_error  '${foo+bar} not bar when foo is not null'
```

Comment:
Is it worth testing the `:+` variation when foo is not null as well?

Reply:
I added a test case for it.